### PR TITLE
Model proxy microservice

### PR DESCRIPTION
**Pull request by atemate**
1. `model-proxy`: a web service exposed to the outside world that receives inference requests, logs them to Postgres (separate from mlflow), calls the model the deployed in another service (still in progress) and returns the result
2. a proper way to do such a logging proxy: to use ELK stack (see for example how [Seldon](https://deploy.seldon.io/docs/getting-started/production-installation/request-logging/) handles it)
3. Exposed as http://model-proxy.lab1-team3.neu.ro

```
# Model inference:
$ curl -X POST -H "Content-Type: application/json" -d '{"text": "awful cafe", "is_positive_user_answered": false}'  http://34.91.75.82/predictions
{"id":2,"text":"awful cafe","is_positive":{"user_answered":false,"model_answered":false},"details":{"mlflow_run_id":"3acade02674549b19044a59186d97db4","inference_elapsed":0.0012352,"timestamp_utc":"2021-02-04T01:23:31.644682"}}

# Get previous inference result:
$ curl http://34.91.75.82/predictions/2
{"text":"awful cafe","is_positive_user_answered":false,"is_positive_model_answered":false,"mlflow_run_id":"3acade02674549b19044a59186d97db4","timestamp_utc":"2021-02-04T01:23:31.714154","id":2}

# Get statistics:
$ curl http://34.91.75.82/statistics   
{"statistics":{"correctness_rate":0.8}}
```

### CODE REVIEW

File: model_proxy/Makefile [hunk 1]
```diff
@@ -0,0 +1,88 @@
+MAKEFILE_PATH = $(abspath $(lastword $(MAKEFILE_LIST)))
+MAKEFILE_DIR = $(dir $(MAKEFILE_PATH))
+
+IMAGE_LOCAL = model_proxy:v1.0
+IMAGE_GCR = gcr.io/mlops-lab1-team3/$(IMAGE_LOCAL)
+
+K8S_NS = mlflow-seldon
```
Comment (atemate) [inline model_proxy/Makefile]: better: `mlflow-model`

File: model_proxy/Makefile [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+k8s-create-postgres-secret: require-POSTGRES_USER require-POSTGRES_PASSWORD
+k8s-create-postgres-secret:
+	kubectl -n $(K8S_NS) create secret generic postgres-config \
+		--from-literal user=$(POSTGRES_USER) \
+		--from-literal password=$(POSTGRES_PASSWORD)
+
+
+.PHONY: k8s-undeploy
+k8s-undeploy:
+	kubectl -n mlflow-seldon delete -f k8s/deployment.yaml
```
Comment (atemate) [inline model_proxy/Makefile]: `-n $(K8S_NS)`

File: model_proxy/app/main.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+
+@app.on_event("shutdown")
+async def shutdown():
+    await database.disconnect()
+
+
+app.include_router(router_ping.router)
+app.include_router(router_predictions.router, prefix="/preds", tags=["predictions"])
+app.include_router(router_statistics.router, prefix="/stats", tags=["statistics"])
```
Comment (atemate) [inline model_proxy/app/main.py]: better full names

File: model_proxy/k8s/deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,75 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: mlflow-seldon-synchronizer
```
Comment (atemate) [inline model_proxy/k8s/deployment.yaml]: change to: `mlflow-model-proxy`

File: model_proxy/Makefile [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+GIT_REPO = https://github.com/dmangonakis/Yelp_Dataset.git
+GIT_BRANCH = $(shell git branch --show-current)
+
+## Run the app
+#
+
+.PHONY: run-app
+run-app:
+	cd $(MAKEFILE_DIR) \
+        && uvicorn app.main:app --host 0.0.0.0 --port 8080 --access-log --use-colors
```
Comment (atemate) [inline model_proxy/Makefile]: just true

File: model_proxy/Makefile [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+	gcloud auth configure-docker
+	docker tag $(IMAGE_LOCAL) $(IMAGE_GCR)
+	docker push $(IMAGE_GCR)
+
+## Deploy the app to k8s
+#
+
+.PHONY: k8s-create-namespace
+k8s-create-namespace:
+	kubectl create namespace $(K8S_NS)
```
Comment (atemate) [inline model_proxy/Makefile]: `|| true`

**Review by Jimaras08 — APPROVED** [review]

File: model_proxy/Makefile [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+.PHONY: k8s-create-postgres-secret
+k8s-create-postgres-secret: require-POSTGRES_USER require-POSTGRES_PASSWORD
+k8s-create-postgres-secret:
+	kubectl -n $(K8S_NS) create secret generic postgres-config \
+		--from-literal user=$(POSTGRES_USER) \
+		--from-literal password=$(POSTGRES_PASSWORD)
+
+
+.PHONY: k8s-undeploy
```
Comment (atemate) [inline model_proxy/Makefile]: `k8s-down-deployment`

File: model_proxy/Makefile [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+k8s-deploy: require-GIT_REPO require-GIT_BRANCH
+k8s-deploy: require-IMAGE_GCR require-POSTGRES_HOST
+k8s-deploy:
+	echo "Deploying from branch origin/$(GIT_BRANCH)"
+	cd $(MAKEFILE_DIR) \
+        && export GIT_REPO=$(GIT_REPO) \
+        && export GIT_BRANCH=$(GIT_BRANCH) \
+        && export IMAGE_GCR=$(IMAGE_GCR) \
+        && export POSTGRES_HOST=$(POSTGRES_HOST) \
+        && envsubst < k8s/deployment.yaml | kubectl -n $(K8S_NS) apply -f -
```
Comment (atemate) [inline model_proxy/Makefile]: use single export

File: model_proxy/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,10 @@
+FROM python:3.7
+
+ENV PYTHONDONTWRITEBYTECODE 1
+ENV PYTHONUNBUFFERED 1
+
+COPY requirements.txt /opt/requirements.txt
+
+RUN pip install -U pip \
```
Comment (atemate) [inline model_proxy/Dockerfile:8]: --no-cache

File: model_proxy/Makefile [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+	kubectl -n $(K8S_NS) create secret generic postgres-config \
+		--from-literal user=$(POSTGRES_USER) \
+		--from-literal password=$(POSTGRES_PASSWORD)
+
+
+.PHONY: k8s-undeploy
+k8s-undeploy:
+	kubectl -n mlflow-seldon delete -f k8s/deployment.yaml
+
+.PHONY: k8s-deploy
```
Comment (atemate) [inline model_proxy/Makefile]: goes before undeploy

File: model_proxy/Makefile (hunk 8 above)
Comment (atemate) [inline model_proxy/Makefile]: envsubst '$A $B' file | ...

File: model_proxy/Makefile [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+## Prepare remote app image
+#
+
+.PHONY: build-image
+build-image:
+	cd $(MAKEFILE_DIR) \
+        && docker build . -t $(IMAGE_LOCAL)
+
+.PHONY: k8s-push-image
+k8s-push-image:
```
Comment (atemate) [inline model_proxy/Makefile]: push-image

**Pull request merged**